elasticsearch/connection/http_urllib3.py

```diff
@@ -230,17 +229,12 @@ def perform_request(
             if not isinstance(method, str):
                 method = method.encode("utf-8")
 
-            request_headers = self.headers
-            if headers:
-                request_headers = request_headers.copy()
-                request_headers.update(headers)
+            request_headers = self.headers.copy()
+            request_headers.update(headers or ())
+
             if self.http_compress and body:
-                try:
-                    body = gzip.compress(body)
-                except AttributeError:
-                    # oops, Python2.7 doesn't have `gzip.compress` let's try
-                    # again
-                    body = gzip.zlib.compress(body)
+                body = self._gzip_compress(body)
```

Comment:
this looks like it might interfere with the calls to `log_request_success` and `log_request_fail` below. Not a new bug if true, but probably worth fixing